File: tests/workers/test_yolov3.py
# Copyright 2021 Xilinx, Inc.
# Copyright 2022 Advanced Micro Devices, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import json
import os

import numpy as np
import pytest
from helper import root_path, run_benchmark

import proteus


@pytest.mark.extensions(["aks", "vitis"])
@pytest.mark.fpgas("DPUCADF8H", 1)
@pytest.mark.usefixtures("load")
class TestInferImageYoloV3DPUCADF8H:
    """
    Test the yolov3 worker
    """

    model = "AksDetect"
    parameters = {
        "aks_graph_name": "yolov3",
        "aks_graph": "${AKS_ROOT}/graph_zoo/graph_yolov3_u200_u250_proteus.json",
    }

    def send_request(self, request, check_asserts=True):
        """
        Sends the given request to the server and asserts common checks

        Args:
            request (InferenceRequest): Request to send to the server
            image (np.ndarray): Golden image to check against
            check_asserts (bool): Verify image against golden

        Returns:
            Response: Response as a dictionary
        """

        try:
            response = self.rest_client.modelInfer(self.endpoint, request)
        except proteus.ConnectionError:
            pytest.fail(
                "Connection to the proteus server ended without response!", False
            )

        num_inputs = len(request.getInputs())
        # for this picture and xmodel combination, we expect the following output with Vitis 2.5
        gold_response_output = [
            1,
            0.9823938,
            200.2854,
            213.84378,
            178.46155,
            107.93164,
            14,
            0.99329937,
            240.7067,
            154.99591,
            90.769196,
            142.7096,
        ]

        if check_asserts:
            assert not response.isError(), response.getError()
            # content = response.json()
            assert response.id == ""
            assert response.model == "yolov3"
            outputs = response.getOutputs()
            assert len(outputs) == num_inputs
            for index, output in enumerate(outputs):
                assert output.name == "input" + str(index)
                assert output.datatype == proteus.DataType.FP32
                assert output.parameters.empty()
                data = output.getFp32Data()
                num_boxes = int(len(data) / 6)
                assert output.shape == [6, num_boxes]
                assert len(data) == len(gold_response_output)
                assert np.allclose(data, gold_response_output, 0.01, 0)
        return response

    def construct_request(self, asTensor):
        image_path = str(root_path / "tests/assets/bicycle-384566_640.jpg")

        batch = 1
        images = [image_path] * batch
        request = proteus.ImageInferenceRequest(images, asTensor)

        return request

    def test_yolov3_dpucadf8h_1(self):
        """
        Send a request to yolov3 as base64-encoded data
        """
        request = self.construct_request(False)
        self.send_request(request)

    @pytest.mark.benchmark(group="yolov3_dpucadf8h")
    def test_benchmark_yolov3_dpucadf8h_1(self, benchmark):
        request = self.construct_request(False)
        options = {
            "model": self.model,
            "parameters": self.parameters,
            "type": "rest (pytest)",
            "config": "N/A",
        }
        run_benchmark(
            benchmark, "yolov3_1", self.rest_client.modelInfer, request, **options
        )

File: tests/workers/xmodel/test_yolov3.cpp
// Copyright 2022 Advanced Micro Devices, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <algorithm>            // for max
#include <array>                // for array
#include <cstdint>              // for uint64_t
#include <initializer_list>     // for initializer_list
#include <memory>               // for unique_ptr, alloca...
#include <opencv2/core.hpp>     // for CV_32FC3
#include <opencv2/imgproc.hpp>  // for COLOR_BGR2RGB
#include <string>               // for string, allocator
#include <vector>               // for vector

#include "amdinfer/amdinfer.hpp"                   // for InferenceRequest
#include "amdinfer/pre_post/image_preprocess.hpp"  // for ImagePreprocessOpt...
#include "amdinfer/testing/get_path_to_asset.hpp"  // for getPathToAsset
#include "amdinfer/testing/gtest_fixtures.hpp"     // for SuiteApiResolver

namespace amdinfer {

using Images = std::vector<std::vector<float>>;

Images preprocess(const std::vector<std::string>& paths) {
  // this model uses images of 416x416
  const int image_size = 416;
  const auto convert_scale = 1 / 255.0;

  amdinfer::pre_post::ImagePreprocessOptions<float, 3> options;
  options.height = image_size;
  options.width = image_size;

  options.resize_algorithm = pre_post::ResizeAlgorithm::CenterCrop;

  options.convert_color = true;
  options.color_code = cv::COLOR_BGR2RGB;

  options.convert_type = true;
  options.type = CV_32FC3;
  options.convert_scale = convert_scale;

  options.normalize = true;
  options.order = amdinfer::pre_post::ImageOrder::NHWC;
  options.mean = {0, 0, 0};
  options.std = {1, 1, 1};
  return pre_post::imagePreprocess(paths, options);
}

void postprocess(const amdinfer::InferenceResponseOutput& output) {
  // TODO(varunsh): needs to be implemented
  (void)output;
}

std::string workerLoad(Client* client, RequestParameters* parameters) {
  return client->workerLoad("Xmodel", parameters);
}

std::vector<InferenceRequest> constructRequests(const Images& images) {
  std::vector<InferenceRequest> requests;
  requests.reserve(images.size());

  const std::initializer_list<uint64_t> shape = {224, 224, 3};

  for (const auto& image : images) {
    requests.emplace_back();
    // NOLINTNEXTLINE(google-readability-casting)
    requests.back().addInputTensor((void*)image.data(), shape, DataType::Int8);
  }

  return requests;
}

void validate(const std::vector<InferenceResponse>& responses) {
  // TODO(varunsh): add post-processing and verify output
  for (const auto& response : responses) {
    auto outputs = response.getOutputs();
    EXPECT_EQ(outputs.size(), 3);
  }
}

void test0(Client* client) {
  if (!serverHasExtension(client, "vitis")) {
    GTEST_SKIP() << "Vitis AI support required from the server but not found";
  }

  if (!client->hasHardware("DPUCADF8H", 1)) {
    GTEST_SKIP() << "At least one DPUCADF8H required on the server";
  }

  const auto test_asset = getPathToAsset("asset_bicycle-384566_640.jpg");
  const auto xmodel = getPathToAsset("u250_yolov3");

  amdinfer::RequestParameters parameters;
  parameters.put("model", xmodel);

  auto images = preprocess({test_asset});
  auto endpoint = workerLoad(client, &parameters);
  auto requests = constructRequests(images);
  auto responses = inferAsyncOrdered(client, endpoint, requests);
  validate(responses);
}

#ifdef AMDINFER_ENABLE_GRPC
// @pytest.mark.extensions(["vitis"])
// @pytest.mark.fpgas("DPUCADF8H", 1)
// NOLINTNEXTLINE(cert-err58-cpp, cppcoreguidelines-owning-memory)
TEST_F(GrpcFixture, WorkersXmodelYolov3) { test0(client_.get()); }
#endif

// @pytest.mark.extensions(["vitis"])
// @pytest.mark.fpgas("DPUCADF8H", 1)
// NOLINTNEXTLINE(cert-err58-cpp, cppcoreguidelines-owning-memory)
TEST_F(BaseFixture, WorkersXmodelYolov3) {
  NativeClient client;
  test0(&client);
}

#ifdef AMDINFER_ENABLE_HTTP
// @pytest.mark.extensions(["vitis"])
// @pytest.mark.fpgas("DPUCADF8H", 1)
// NOLINTNEXTLINE(cert-err58-cpp, cppcoreguidelines-owning-memory)
TEST_F(HttpFixture, WorkersXmodelYolov3) { test0(client_.get()); }
#endif

}  // namespace amdinfer
